gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupGlobalStep.java

```diff
@@ -89,6 +92,17 @@ public ElementRequirement getMaxRequirement() {
 
     @Override
     protected Traverser.Admin<S> processNextStart() {
+        if (null != this.barrier) {
+            this.barrierIterator = this.barrier.entrySet().iterator();
+            this.barrier = null;
+        }
+        while (this.barrierIterator != null && this.barrierIterator.hasNext()) {
+            if (null == this.barrierIterator)
```

Comment:
`this.barrierIterator` can never be null within within the `while()` loop. Unless I overlooked something fundamental, `processNextStart` can be simplified to:

```
protected Traverser.Admin<S> processNextStart() {
    if (null != this.barrier) {
        this.barrierIterator = this.barrier.entrySet().iterator();
        this.barrier = null;
        while (this.barrierIterator.hasNext()) {
            final Map.Entry<Object, Traverser.Admin<S>> entry = this.barrierIterator.next();
            if (this.duplicateSet.add(entry.getKey()))
                return PathProcessor.processTraverserPathLabels(entry.getValue(), this.keepLabels);
        }
    }
    return PathProcessor.processTraverserPathLabels(super.processNextStart(), this.keepLabels);
}
```